runtime/src/iree/hal/drivers/hip/dispatch_thread.c

```diff
@@ -0,0 +1,174 @@
+  return 0;
+}
+
+iree_status_t iree_hal_hip_dispatch_thread_initialize(
+    iree_allocator_t host_allocator,
+    iree_hal_hip_dispatch_thread_t** out_thread) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  *out_thread = NULL;
+  iree_hal_hip_dispatch_thread_t* thread = NULL;
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0,
+      iree_allocator_malloc(host_allocator, sizeof(*thread), (void**)&thread));
+
+  thread->do_exit = false;
+  iree_slim_mutex_initialize(&thread->mutex);
+  iree_hal_hip_dispatch_queue_initialize(host_allocator, &thread->queue);
+  thread->failure_status = iree_ok_status();
+  thread->host_allocator = host_allocator;
+  iree_notification_initialize(&thread->notification);
+
+  iree_thread_create_params_t params;
+  memset(&params, 0x00, sizeof(params));
+  params.name = iree_make_cstring_view("iree-hal-hip-dispatch");
+  iree_status_t status =
+      iree_thread_create((iree_thread_entry_t)iree_hal_hip_dispatch_thread_main,
+                         thread, params, host_allocator, &thread->thread);
+  if (!iree_status_is_ok(status)) {
+    iree_hal_hip_dispatch_queue_deinitialize(&thread->queue);
+    iree_slim_mutex_deinitialize(&thread->mutex);
+    iree_allocator_free(host_allocator, thread);
+  }
+  IREE_TRACE_ZONE_END(z0);
+  *out_thread = thread;
+  return status;
+}
+
+void iree_hal_hip_dispatch_thread_deinitialize(
+    iree_hal_hip_dispatch_thread_t* thread) {
+  if (!thread) {
+    return;
+  }
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_slim_mutex_lock(&thread->mutex);
+  thread->do_exit = true;
+  iree_slim_mutex_unlock(&thread->mutex);
+
+  iree_notification_post(&thread->notification, IREE_ALL_WAITERS);
+  // There is only one owner for the thread, so this also joins the thread.
+  iree_thread_release(thread->thread);
+
+  iree_status_ignore(thread->failure_status);
+  iree_hal_hip_dispatch_queue_deinitialize(&thread->queue);
+  iree_slim_mutex_deinitialize(&thread->mutex);
+  iree_allocator_free(thread->host_allocator, thread);
+  IREE_TRACE_ZONE_END(z0);
+}
+
+iree_status_t iree_hal_hip_dispatch_thread_add_dispatch(
+    iree_hal_hip_dispatch_thread_t* thread,
+    iree_hal_hip_dispatch_callback_t dispatch, void* user_data) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  iree_slim_mutex_lock(&thread->mutex);
+  iree_status_t status = iree_status_clone(thread->failure_status);
+
+  iree_hal_hip_dispatch_thread_dispatch_t dispatch_data = {
+      .dispatch = dispatch,
+      .user_data = user_data,
+  };
+
+  if (iree_status_is_ok(status)) {
+    status =
+        iree_hal_hip_dispatch_queue_push_back(&thread->queue, dispatch_data);
+  }
+  if (!iree_status_is_ok(status)) {
+    iree_status_ignore(thread->failure_status);
+    thread->failure_status = iree_status_clone(status);
+  }
+  iree_slim_mutex_unlock(&thread->mutex);
+  iree_notification_post(&thread->notification, IREE_ALL_WAITERS);
+
+  IREE_TRACE_ZONE_END(z0);
+
+  if (!iree_status_is_ok(status)) {
+    iree_status_ignore(dispatch(user_data, iree_status_clone(status)));
```

Comment:
this should be in the trace zone so it's accounted for - otherwise you'll see any zones nested within as nested in the caller and it'll be hard to tell where they are coming from